### PR TITLE
[Bugfix][Model Runner V2][Spec Decode] Fix off-by-one in bad_words draft-prefix matching

### PR DESCRIPTION
**Pull request by jyan-R**
## Purpose

Fix an off-by-one in the spec-decode branch of `_bad_words_kernel` (`vllm/v1/worker/gpu/sample/bad_words.py`), present since the kernel was introduced in #33433.

The sampler passes `input_ids` gathered at `logits_indices`, so for a spec-decode request the per-request local layout is: local position 0 = last committed sampled token, local position j = draft token d_{j-1}. The sibling kernels index draft tokens accordingly (`_penalties_kernel` uses `start_idx + prev_pos + 1`; thinking budget's `_load_effective_token` uses `+ 1` with an explicit comment). `_bad_words_kernel` read `input_ids_ptr + cur_req_first_pos + spec_offset` without the `+ 1`, so every draft-region read was shifted back by one token (the boundary token was read twice).

Consequences when bad_words and speculative decoding are combined and a multi-token bad word's prefix window reaches into the draft tokens:
- prefix matches ending in the draft region are missed (the banned token can be sampled/accepted);
- matches spanning the committed/draft boundary fire one verification row late;
- the duplicated boundary read can spuriously match and ban an innocent token.

Single-token bad words and non-spec rows are unaffected. Model Runner V1 has a separate implementation and is unaffected.

**Not a duplicate:** searched open PRs/issues for `bad_words` + spec decode / off-by-one / `_bad_words_kernel`; no existing fix found (closest is #34213, which only touches penalties).

## Test Plan

New unit test (first coverage for this kernel), following the pattern of `tests/v1/worker/test_gpu_thinking_budget.py`:

```bash
pytest -v tests/v1/worker/test_gpu_bad_words.py
```

Scenario: committed output `[10, 11]`, draft tokens `[12, 13]`. The core regression case is the boundary-spanning prefix (`[11, 12, 30]` must mask at the row where the prefix completes, not one row later); the other three cases pin the remaining coverage axes: a positive match at draft offset >= 1 (`[12, 13, 40]`), a no-double-count guard for the boundary token (`[11, 11, 50]`), and a committed-only baseline (`[10, 11, 60]`). Assertions compare the full logits tensor.

## Test Result

Ran on an A100 SXM4 80GB (torch 2.13.0+cu130, triton 3.7.1, Python 3.11), with the branch installed editable via `VLLM_USE_PRECOMPILED` (nightly x86_64 wheel `66728feb1`):

- **With the fix (HEAD): `4 passed`.**
- **With the pre-fix kernel** (parent commit's `bad_words.py` swapped in): **`3 failed, 1 passed`** — exactly the three spec-branch cases fail; the committed-only baseline passes.
- `ruff check` / `ruff format --check` pass on both changed files.
- Cross-validation: the kernel's scalar loop was transcribed to pure Python and checked against an independent reference matcher over 20,000 randomized spec-decode scenarios; with the `+ 1` the kernel matches the reference in all cases, without it ~24% of scenarios diverge.

AI assistance was used for this change (Claude); I have reviewed every changed line.

### CODE REVIEW

File: tests/v1/worker/test_gpu_bad_words.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        prompt_len=PROMPT_LEN,
+        all_token_ids=COMMITTED,
+        num_computed_tokens=len(COMMITTED),
+        max_tokens=32,
+    )
+    req_states.apply_staged_writes()
+
+    state = BadWordsState(req_states)
+    state.add_request(3, SamplingParams(_bad_words_token_ids=bad_words_token_ids))
+    state.apply_staged_writes()
```
[the remark this answers is not in the document]
Comment (jyan-R) [reply tests/v1/worker/test_gpu_bad_words.py:52]: Done in 53b989f39 — the request index is now derived from `req_states.req_id_to_index["req"]` and threaded through `_make_state`/`_apply`, covering both flagged lines.

**Review by njhill — APPROVED** [review]
Thanks @jyan-R

**Comment by jyan-R** [comment]
CI status note: the remaining red job `Multi-Modal Models (Standard 4, other + whisper)` failed twice (initial + retry) with an identical, PR-unrelated signature — gemma4 engine-core warmup crashing on a PyTorch-internal NVML assert (`CUDACachingAllocator.cpp:1407`), preceded by stale `torch_compile_cache` cubin misses on the node. The AMD mirror of the same suite passed in this build, and the jobs covering this PR's area (`v1-sample-plus-logits`, `v1-others-cpu`) are green. Filed #52403 to track the infra failure; not retrying further to avoid burning CI.

**Comment by jyan-R** [comment]
Build #83984 finished. Final tally: everything green except (a) `cpu-language-generation-and-pooling-model-tests` — **Expired**, never picked up by an agent, so it has not actually run; and (b) the whisper job tracked in #52403. Issuing one `/ci retry` to requeue the expired job; the whisper job is expected to stay red unless the node's stale compile cache cleared (see #52403).
🤖 Posted by [Claude Code](https://claude.com/claude-code) on behalf of [@jyan-R](https://github.com/jyan-R)
